hudi-client/hudi-client-common/src/main/java/org/apache/hudi/async/HoodieAsyncTableService.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hudi.async;
+
+import org.apache.hudi.client.RunsTableService;
+import org.apache.hudi.config.HoodieWriteConfig;
+
+import java.util.function.Function;
+
+public abstract class HoodieAsyncTableService extends HoodieAsyncService implements RunsTableService {
+
+  protected HoodieWriteConfig writeConfig;
+
+  protected HoodieAsyncTableService() {
+  }
+
+  protected HoodieAsyncTableService(HoodieWriteConfig writeConfig) {
+    this.writeConfig = writeConfig;
+  }
+
+  protected HoodieAsyncTableService(HoodieWriteConfig writeConfig, boolean runInDaemonMode) {
+    super(runInDaemonMode);
+    this.writeConfig = writeConfig;
+  }
+
+  @Override
+  public void start(Function<Boolean, Boolean> onShutdownCallback) {
+    if (!tableServicesEnabled(writeConfig)) {
```

Comment:
also, trying to understand where exactly this guard need to be added. 
As per master, this is my understanding.
we check the configs and may not start the async service only.
 for eg
```
if (writeClient.getConfig().isAutoClean() && writeClient.getConfig().isAsyncClean()) {
      asyncCleanerService = new AsyncCleanerService(writeClient);
      asyncCleanerService.start(null);
    } else {
      LOG.info("Async auto cleaning is not enabled. Not running cleaner now");
    }
```

So, that the shutdown code looks as below.
```
    if (asyncCleanerService != null) {
      LOG.info("Waiting for async cleaner to finish");
      try {
        asyncCleanerService.waitForShutdown();
      } catch (Exception e) {
        throw new HoodieException("Error waiting for async cleaning to finish", e);
      }
    }
```
```
  public static void forceShutdown(AsyncCleanerService asyncCleanerService) {
    if (asyncCleanerService != null) {
      LOG.info("Shutting down async cleaner");
      asyncCleanerService.shutdown(true);
    }
  }
```

but now as per this patch, I feel this is an additional check we are adding which may not be invoked at all since the async service should not be instantiated at all if the config of interest is not enabled only.
Or how are we ensuring HoodieAsyncService.shutdown() and HoodieAsyncService.waitForShutdown accounts for tableServiceDisabled flow.

Reply:
the main idea is: we can't prevent an async service to be instantiated but we can prevent them from running by checking this in `start()`. It'd be some code smell if we always check configs before creating an async service; we better encapsulate the checking inside the services, hence this patch's approach. Since the start() will never create a future to run async, shutdown() or waitForShutdown() should just exit no-op. From the existing codebase, i see `shutdown()` null-check `executor` but `waitForShutdown()` does not null-check `future`. So I should add a null-check there as well.

Reply:
sounds good.